---
title: "Genotype Matrixes"
author: "FiG-T"
data: "`r Sys.Date()`"
output: 
  html_document: 
    keep_md: yes
    toc: yes
toc: TRUE
editor_options:
  markdown:
    wrap: 80
  chunk_output_type: inline
---

## Extracting Genotype Matrices

The Generalised Linear Models require a the genetic information to be stored as a list of genotypes (1 or 0 for each SNP). This can be extracted from the filtered VCF file.

### Import the Filtered VCF file

```{r import_vcf}

vcf <- vcfR::read.vcfR(
  file = "data/vcf/snp_sites_filtered_vcf_maf0.005_noRef_092023.recode.vcf", 
  # this file is the same as the one used for vcf processing
  verbose = TRUE
)

```

### Extracting the genotype matrix

```{r extract_gt_matrix}

# extract the Genotype matrix
gt <- vcfR::extract.gt(
  x = vcf, 
  element = "GT"
)

gt_t <- t(gt) # transpose so that alleles are the columns not rows

# convert to a tibble 
gt_t <- tidyr::as_tibble(
  gt_t,
  rownames = "id" # retain the rownames (ids) and place into a column. 
)
```

The resulting tibble has the genotype data for all individuals that passed the
filtering steps (23,790).

```{r save_GT_table}

feather::write_feather(
  x = gt_t, 
  "data/gt_table_02_2024.feather"
)

```

## Creating a genlight object

```{r create_genlight, echo=TRUE}

# this can be done using the vcfR package:
gl <- vcfR::vcfR2genlight(
  x = vcf
) # this will omit any non-biallelic sites
```

## Generating PCA scores

```{r PCA_scores, echo = FALSE, include = FALSE}

pca <- adegenet::glPca(
  #x = vcf.gl,  # FAILSAFE-remove initial hashtag
  nf = 4,          # how many PCs to retain
  parallel = TRUE, # unless specified otherwise, the max number of available cores will be used. 
)

```

```{r save_pca_scores}
# convert to a tibble
pca_scores <- tidyr::as_tibble(pca$scores)
pca_eig <- tidyr::as_tibble(pca$eig)
pca_loadings <- tidyr::as_tibble(pca$loadings)

pca_list <- list(pca_scores, pca_eig, pca_loadings)

# as the pca takes a long time to run ... save these scores... 
feather::write_feather(
  x = pca_scores, 
  path = "data/feather/pca_scores_0.005_03_2024.feather"
)
feather::write_feather(
  x = pca_eig, 
  path = "data/feather/pca_eig_0.005_03_2024.feather"
)
feather::write_feather(
  x = pca_loadings, 
  path = "data/feather/pca_loadings_0.005_03_2024.feather"
)
```

To calculate the percentage of variance explained by each PC.

```{r pca_variance}
# calculate the sum of all eigenvalues
eig.total <- sum(pca$eig)

pca_var <- c()
for (i in 1:4){
  
  pca_var[i] = formatC(
      head(pca$eig)[i]/eig.total * 100
    )
  
 # where i is the number of the PC you wish to show...
}

# variance explained: 
# "19.09" "11.77" "7.153" "5.766"

```
